// @node-red/flow-parser/lib/node.js
/* node */
const utils = require('./utils.js');

/** Class for FMNode (incl. subflow instances) */
class FMNode {
    /**
     * constructor
     * @param {string} id - node id.
     */
    constructor(id) {
        this.id = id || utils.genId();
        this.type = "";
        this.x = 0;
        this.y = 0;
        this.wires = [];  // wires[portNum] = [nextId]
    }

    /**
     * connect two nodes
     * @param {number} fromport - node port number of this node
     * @param {FMNode} nextnode - reference to next node
     * @returns {boolean} - true if success
     */
    connectTo(fromport, nextnode) {
        if (this.wires.length > fromport && 'id' in nextnode) {
            this.wires[fromport].push(nextnode.id);
            return true;
        } else {
            return false;
        }
    }


    /**
     * disconnect designated node with the node
     * @param {number} fromport - portnumber. if -1, remove all occurrence of nextnode
     * @param {FMNode} nextnode - node. if null, remove all node on the port.
     * @returns {boolean} - true if success
     */
    disconnectTo(fromport, nextnode) {
        if (fromport == -1) {
            for (let i = 0; i < this.wires.length; i++) {
                if (nextnode) {
                    this.wires[i] = this.wires[i].filter(w => w !== nextnode.id);
                } else {
                    this.wires[i] = [];
                }
            }
            return true;
        } else if (this.wires.length > fromport) {
            this.wires[fromport] = this.wires[fromport].filter(w => w !== nextnode.id);
            return true;
        } else {
            return false;
        }
    }
}

module.exports = FMNode;
