// nrlint-plugin-core/lib/rules/flow-size.js
/**
 * Check size of each flow
 * @param {FlowSet} afs complete flow set
 * @param {*} conf configuration for this rule. {name: "flowsize", maxSize: number, severity: "warn" }
 * @param {*} cxt context
 */
function checkFlowSize(afs, conf, cxt) {
    const flows = afs.flows;

    flows.forEach((flow) => {
        const nodes = flow.nodes;
        let maxSize = 100;
        if ('maxSize' in conf) {
            maxSize = conf.maxSize;
        }
        if (nodes.length > maxSize) {
            cxt.report({
                location: [flow.id],
                message: "flow too large"
            })
        }
    });
}

module.exports = {
    meta: {
        type: "suggestion",
        severity: "warn",
        docs: {
            description: "limit the number of nodes on any one flow"
        },
        schema: [
            {
                type: "object",
                properties: {
                    maxSize: { type: "number" }
                }
            }
        ]
    },
    check: checkFlowSize
};
